_posts/2023-12-20-standard-libraries.markdown

```diff
@@ -0,0 +1,407 @@
+* The `FromUTF8Checked` function takes in a `seq<uint8>`, where `uint8` is a `newtype` definition from `Std.BoundedInts`.
+  This library defines lots of common fixed-bit-width integer types that the Dafny compiler will map to
+  the native integer types of the target language for improved efficiency and memory usage.
+* Because `bv8` is a distinct type from `uint8`, we have to explicitly convert between them.
+  This is a side effect of the Dafny standard libraries having multiple independent contributions from various
+  projects, and you can expect future versions of Dafny to provide more versions of common utilities
+  to make combining things a bit smoother.
+* `FromUTF8Checked` returns an `Option<string>`, but `FileIO.ReadBytesFrom` returns a `Result<seq<bv8>, string>`.
+  To unify with a common result type, we use `Option.ToResult` to convert the former value to a `Result`,
+  which is a handy and common trick. See [the documentation for `Std.Wrappers`](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyStandardLibraries/src/Std/Wrappers.md) for more details.
+
+## Sprinting to the Finish
+
+Now let's put it all together and write our main method:
+
+```dafny
+method Main() {
+  var input :- expect ReadPuzzleInput();
+
+  var lines := Seq.Split(s, '\n');
+
+  var calibrationValues :- expect Seq.MapWithResult(CalibrationValue, lines);
+
+  var total := Seq.FoldLeft((x, y) => x + y, 0, calibrationValues);
+  print total, "\n";
+}
+```
+
+We've used one more trick with failure-compatible types in this method:
+`:- expect`, a variation on the elephant operator to make "assign or halt" statements.
+`ReadPuzzleInput` returns a `Result`, but if it fails there's nothing more we can do
+in the main method than printing an error and exiting.
+`:- expect` means that if the right-hand side of the statement is a failure,
+Dafny will immediately halt and print the failure value to the console.
+This is particularly valuable for writing tests in Dafny.
+Note also that the `Seq` library continues to work hard for us,
+splitting the input into lines,
+mapping our `CalibrationValue` function over the sequence of lines,
+and summing the results into the final answer.
+`MapWithResult` is a short-circuiting variation of `Map` where the mapped function can fail.
+
+If we put the sample input from the puzzle description into `input.txt`
+(not the actual puzzle input - I draw the line at publishing the answer in this blog post,
+you're just going to have to run the solution yourself to get that!),
+we can run our project and get...
+
+```dafny
+% dafny run dfyconfig.toml
+
+Dafny program verifier finished with 3 verified, 0 errors
+142
+```
+
+Huzzah! We have a working solution to the first puzzle in less than 40 lines of Dafny code
+(see below for the complete program),
+all thanks to the Dafny standard libraries.
+
+```dafny
```

Comment:
Can you use import statements like John's blog post, so that you can verify this piece of code?

Reply:
I'm still trying to figure out the best way to do that and get nice syntax highlighting on the snippets. I'm not using Madoko and trying to find a more standard solution, hence experimenting with pandoc. My fallback approach will be to use the `verification-compelling-verification-steps.js` script and the `{% include ... %}` directives it supports, as the most recent post did.

Reply:
I ended up switching to Madoko and just including the final solution source, so at least that is continuously checked. It didn't seem worth it to convert the other snippets to source since they aren't complicated, and some are not even expected to be valid since they are partial work-in-progress.